sqlglot/parser.py

```diff
@@ -4199,7 +4199,10 @@ def _parse_type(
 
                 return self.expression(exp.Cast, this=this, to=data_type)
 
-            if data_type.expressions:
+            # We may have set `expressions` using the `TYPE_CONVERTERS` mapping, in which
+            # case we shouldn't really move on with the data type, but instead retreat to
```

Comment:
i don't really understand this comment, what does it mean set the expressions using type converters, and how does the index difference detect this

Reply:
Fair enough.

The gist of it is that `expressions` gets set by the *parser* (i.e. *not* because of a converter) when you have something like `DECIMAL(38, 0)` in the input SQL. In that case, you'll have these tokens: `[DECIMAL, L_PAREN, NUMBER, COMMA, NUMBER, R_PAREN]`.

If the index difference in the condition is greater than 1, that means the parser itself must have consumed additional tokens such as the `DECIMAL`'s scale and precision here.

If it's not greater than 1, then it must be 1 (because we parsed a type in L4189), meaning that the `expressions` arg of the `DataType` got set by a type converter (e.g. Snowflake converts `DECIMAL` to `DECIMAL(38, 0)`).

Reply:
why would the expressions get set if the index is 1?

Reply:
Because `DECIMAL` in Snowflake gets converted into an AST node that has `expressions` set to `(38, 0)` ([context](https://github.com/tobymao/sqlglot/blob/main/sqlglot/dialects/dialect.py#L1166-L1176)), to facilitate transpilation.

Reply:
let's add that to the comment

Reply:
Done: https://github.com/tobymao/sqlglot/pull/3566/commits/ed01a1c4d60956ca39c52e77ad866f2668eacdf6